(WIP) Feature: Add guided example for reusable pipelines (osft-minimal) (#31)
red-hat-data-services/red-hat-ai-examples: examples/fine-tuning/pipelines/training-hub/README.md

# Fine-tuning pipelines on Red Hat OpenShift AI (RHOAI)

This guided example goes through how to set up the Training Hub's fine-tuning pipeline on Red Hat OpenShift AI.

Our example will go through: dataset download, training on a single node with GPU, evaluation of the fine-tuned model, and registering the model to Model Registry with use of the OSFT algorithm.

## Note

This example is compatible with RHOAI version 3.3

## Pipelines Overview

The pipelines are built using modular, reusable KFP components that can be integrated into your own custom pipelines.

There are four flavours of the pipelines:

> - **OSFT** - OSFT pipeline introduces expert level pipeline which provides a huge set of parameters to fine grain OSFT based fine tuning.

Reviewer: Does this sound better? Same for other pipelines
```suggestion
- **OSFT** - Expert-level pipeline with extensive parameters for fine-grained control over OSFT-based fine-tuning.
```

Author: Yes, that reads better, thanks :+1:

- **OSFT-minimal** - OSFT-minimal pipeline introduces entry level pipeline which can be used to fine tune model with as little as just one required parameter.
- **SFT** - SFT pipeline introduces expert level pipeline which provides a huge set of parameters to fine grain SFT based fine tuning.
- **SFT-minimal** - SFT-minimal pipeline introduces entry level pipeline which can be used to fine tune model with as little as just one required parameter.

Each of the pipelines consists of **four** components:

> - **Dataset Download** - Downloads and validates the dataset, ensuring it's in the expected chat format for Training Hub

Reviewer 2: @MStokluska Should we link to the components here?

Author: added in latest commit

- **oci-pull-secret-model-download (required for private OCI base models)**
  - **Used by**: Training step when `phase_02_train_man_train_model` is an `oci://` reference.
  - **Key**:
    - `OCI_PULL_SECRET_MODEL_DOWNLOAD` – the contents of a Docker `config.json`, used by `skopeo` to authenticate to the registry.

## Pipelines Parameters

Each pipeline flavor exposes different set of parameters. With minimal pipelines keeping the number of parameters to absolute minimum, the expert pipelines
expose every param that underlying components support:

- **OSFT** - [list of params](https://github.com/red-hat-data-services/pipelines-components/blob/main/pipelines/training/osft/README.md#inputs-)
- **OSFT-minimal** - [list of params](https://github.com/red-hat-data-services/pipelines-components/blob/main/pipelines/training/osft_minimal/README.md#inputs-)
- **SFT** - [list of params](https://github.com/red-hat-data-services/pipelines-components/blob/main/pipelines/training/sft/README.md#inputs-)
- **SFT-minimal** - [list of params](https://github.com/red-hat-data-services/pipelines-components/blob/main/pipelines/training/sft_minimal/README.md#inputs-)

## Quick Start

> As part of quick start OSFT-minimal pipeline was used.

Reviewer: ```suggestion
This quick start guide uses the OSFT-minimal pipeline.
```

The only differences between the flavors of pipelines are:

- parameters
- hardware requirements

### 1. Generate the Pipeline YAML

Before we upload the pipelines to RHOAI platform, we need to ensure that the pipelines are in .yaml format.
The compiled pipeline YAML is not included in the repository and must be generated locally, reason being is that some values - like the `storage class`, will
most likely need to be tweaked to the storage class available on cluster.

**Requirements:**

- Python 3.11+
- KFP SDK version **2.15.2** (important: other versions may produce incompatible YAML)

```bash
# Clone the repository

> git clone https://github.com/kubeflow/pipelines-components.git

Reviewer: Good point, @MStokluska could you update this please?

   ![](./docs/03.png)

### 4. Run the Pipeline

1. Navigate to **Develop & train** > **Pipelines** > **Runs** and click **Create run**

   ![](./docs/04.png)

2. Set the following:

   - Experiment - optional - can be left as Default (learn more about how to manage experiments in [official documentation](https://docs.redhat.com/en/documentation/red_hat_openshift_ai_self-managed/3.0/html/working_with_ai_pipelines/managing-pipeline-experiments_ai-pipelines#overview-of-pipeline-experiments_ai-pipelines))
   - Name - required - name of the run
   - Description - optional - description of the run
   - Pipeline - required - select the pipeline name defined in pipeline definition step
   - Pipeline version - required - autoselects latest version

   ![](./docs/05.png)

3. Scroll down to **Parameters** and set the following required parameter:

>    - **phase_01_dataset_man_data_uri** to **LipengCS/Table-GPT:All**

Reviewer: not relevant

Each pipeline step is performed as a separate Kubernetes pod. Pods can be investigated for further details if required.

### 5. Model Registry

1. To view model registered in model registry navigate to **AI hub** > **Registry**

   ![](./docs/10.png)

2. Select the model and navigate to **Versions**, from where you can select your fine-tuned model version and view
metrics in **Details**

> - Note: Version, model name, and registry address are all part of exposed optional parameters in the pipeline.
> - Note: The amount of metrics varies based on the number of evaluation tasks selected; by default, evaluation performs only the arc_easy task.

![](./docs/11.png)

## Building Custom Pipelines with reusable components

The training component is the core of the fine-tuning pipeline. It handles:

- Downloading the base model from Hugging Face or OCI registries
- Configuring and submitting Kubernetes training jobs via the Training Operator
- Monitoring job progress and collecting training metrics
- Producing loss charts and training artifacts

The component is available as a standalone KFP component that you can import and use in your own pipelines:

```python
from kfp import dsl

> from components.training.finetuning import train_model

Reviewer: incorrect
